enhancements/api-review/new-CRD-ManagedClusterVersion.md

```diff
@@ -0,0 +1,490 @@
+  - https://issues.redhat.com//browse/HIVE-2366
+see-also:
+  - n/a
+replaces:
+  - n/a
+superseded-by:
+  - n/a
+---
+
+# ManagedClusterVersion CRD
+
+## Summary
+
+Introduce a new namespaced openshift/api CustomResourceDefinition,
+`ManagedClusterVersion`, as a vehicle for managers of OpenShift fleets to
+expose version and upgrade information in the management cluster.
+
+## Motivation
+
+OpenShift fleet management software is stratified. For example, ACM (Advanced
+Cluster Manager) and OCM (OpenShift Cluster Manager, the front end for the
+Managed OpenShift product) both provide users access and visibility into
+clusters which may internally be controlled by different pieces of fleet
+management software such as [hive](https://github.com/openshift/hive/) and
+[hypershift](https://github.com/openshift/hypershift/). In this picture, hive
+and hypershift interact directly with the "spoke" clusters, whereas ACM and OCM
+rely on data visible only on the "hub" (the management cluster on which the
+hive/hypershift/ACM/OCM software is running). Thus, any spoke-specific
+information \*CM requires to operate must be brokered by the hive/hypershift
+layer.
+
+These different levels may have different responsibilities. Salient to this
+enhancement, the \*CM layer is responsible for driving upgrades. To do so,
+it needs visibility into the spoke cluster's ClusterVersion data. Today the
+only mechanisms available for accessing this information entail logging into,
+or running an agent in, the spoke cluster. This is not ideal:
+* The \*CM software is often subject to security boundaries that make such
+  access difficult or impossible, mechanically and/or by policy strictures.
+* The hive/hypershift layer is already responsible for such communication
+  with the spoke cluster. Introducing the same into the \*CM layer would
+  seem a duplication of effort and an overlap of domain.
+
+This enhancement enables the hive/hypershift layer to copy ClusterVersion
+status information for each of its spokes into a ManagedClusterVersion object
+on the hub. There should be exactly one ManagedClusterVersion instance
+associated with each spoke.
+
+This information can then be consumed/exposed by the \*CM layer to inform
+upgrade decisions driven by/through that layer.
+
+### User Stories
+
+* As a (human or programmatic) manager of a fleet of OpenShift clusters,
+  potentially at disparate versions and/or on disparate infrastructures, I
+  want to have visibility into version and upgrade information of all clusters
+  in my fleet, so that I can appropriately coordinate upgrades.
+* As a programmatic manager of a fleet of OpenShift clusters (i.e. ACM/OCM), I
+  want a common way to view version and upgrade information, regardless of the
+  software layer between me and the spokes, so that I can simplify my code,
+  reduce my test surface, and spend less on maintenance.
+* As a Site Reliability Engineer (SRE) I want to get the recommended version
```

Comment:
Does it make more sense to develop a tool for SRE that sits outside of a cluster and scans clusters instead of running the agent in every cluster?

Reply:
That's actually exactly what this proposal is all about. Hive and hypershift are exactly such tools today: they sit on a hub cluster and collect data from the spokes*. This proposal is about adding ClusterVersion data to what is collected, and doing it in a CRD that both hive and hypershift (and others) can share.

*Though TBH I don't know whether hypershift does it via an in-cluster agent that reports back to the hub. Hive for sure does not -- the controller on the hub polls spoke clusters via clients constructed from admin kubeconfigs.